Make this pipeline configurable

leoisl opened the pull request:
This PR makes this pipeline configurable through a `config.yaml` file instead of needing a very specific directory structure to run it. The `config.yaml` allows you to feed the input tree and the several batches, and to choose where to put the output and which stats to calculate/output (asm, pre, post). See the example for the default `config.yaml` that runs on the test data:

```
########################################################################################################################
# input data
input_tree: .test/resources/test_run.nw

batches:
  test_run:  # batch name
      GCGS0001: .test/resources/isolates/GCGS0001.fa  # sample name and sample path
      GCGS0002: .test/resources/isolates/GCGS0002.fa
      GCGS0003: .test/resources/isolates/GCGS0003.fa
      GCGS0004: .test/resources/isolates/GCGS0004.fa
      GCGS0005: .test/resources/isolates/GCGS0005.fa
  # more batches allowed here
########################################################################################################################


########################################################################################################################
# output configuration
output_dir: .test/results
generate_assembly_stats: True  # True or False
generate_prepropagation_stats: True  # True or False
generate_postpropagation_stats: True  # True or False
########################################################################################################################
```

It also does some other minor misc stuff: clean up, some minor updates to ignore files, etc.

Importantly, I just tested this in the test data and on a single batch. If you could further test this better before merging, would be great!

karel-brinda commented:
Hi Leandro,

thank you for the update!!

One remark – it's important that the input is provided by users in the form of one or more files of files.

These are most typically provided by listing of the input files if interest by `find` or another tool like this. The same for the input tree.

The configuration file should affect only mode of computation, but no modifications should be required to use the pipeline.

Thanks,
Karel

karel-brinda commented:
For instance, the way how I use it is that I clone the repository just once, create FOFs for all datasets of interest, and then let it run.

leoisl commented:
Ok, I've changed the pipeline to how the input was done before, with a very lengthy description on how the input folder should be structured in `config.yaml`:
```
########################################################################################################################
# input configuration
# This parameter denotes the input dir, a path to a directory with a very specific structure, that contains the input
#     data to this pipeline, i.e. the batches and the initial tree for each batch.
# Here we describe this structure in details.
# Each batch named <batch> is composed of two files: a <batch>.txt file and a <batch>.nw file.
#   * The <batch.txt> file is a text file where each line is a path to an assembly file
#      * An assembly file is a file with extensions ["fa", "fasta", "fna", "ffa"], optionally compressed with gzip
#          ("gz" extension) denoting the assembly of a sample;
#      * See .test/resources/test_run.txt for an example of a <batch.txt> file;
#   * The <batch>.nw file is a newick file representing the initial tree of the samples in its respective <batch>.txt
#          file.
#      * See .test/resources/test_run.txt for an example of a <batch>.nw
#   * These two files for each batch must be at the top level of the given directory.
input_dir: resources
########################################################################################################################
```

karel-brinda commented:
Hi Leandro, I'm trying to incorporate the pull request. Two questions:

1. Where exactly is the main config file?
2. How to fix the following error?
```
```

leoisl commented:
> * Where exactly is the main config file?

You can see an example of the config file in `.test/config.yaml`

> How to fix the following error?

Which make command did you run to cause this error?

leoisl commented:
Either that or snakemake version... I have:
```
snakemake --version
7.25.4
```

karel-brinda commented:
Missing config:

```
$ make
snakemake -j --use-conda -p --rerun-incomplete
WorkflowError in file /Users/karel/github/my/mof-compress/workflow/rules/common.smk, line 7:
Workflow defines configfile config.yaml but it is not present or accessible (full checked path: /Users/karel/github/my/mof-compress/config.yaml).
  File "/Users/karel/github/my/mof-compress/workflow/Snakefile", line 16, in <module>
  File "/Users/karel/github/my/mof-compress/workflow/rules/common.smk", line 7, in <module>
```


Confirm that `.test/config.yaml` exists

karel-brinda commented:
```
$ snakemake --version
7.28.1
```

leoisl commented:
just `make` will indeed cause missing config, as there is no `config.yaml` in the root. `make test` switch to the `.test` dir, where the is a `config.yaml`. User is supposed to create their own `config.yaml` based on `.test/config.yaml`. Going back to your previous error:
```
        odir=$(dirname "results/tree/test_run.nw_dirty")
        r=$(realpath --relative-to="$odir" "resources/test_run.nw")
        ln -sf "$r" results/tree/test_run.nw_dirty
        
realpath: illegal option -- -
```

it seems to be an error when running `realpath`. Could you please output the value of `$odir`?

leoisl commented:
I have a cross-platform solution, just a sec

karel-brinda commented:
Regarding make: I think this is reasonable to expect from the user (the pipeline is to be invoked by make anyway, so the user must have it).

leoisl commented:
Added `macos-latest` to the CI check

leoisl commented:
> Regarding make: I think this is reasonable to expect from the user (the pipeline is to be invoked by make anyway, so the user must have it).

yeah, weirdly the GH actions ubuntu worker did not have `make` installed... maybe it does not come by default with ubuntu? ... or for whatever other reason, `prophyle` complained it could not find make...

leoisl commented:
> Added `macos-latest` to the CI check

Just reverted this, mac support is not there yet I guess: https://github.com/karel-brinda/mof-compress/actions/runs/5579387979/jobs/10194859289?pr=31

leoisl commented:
Hey @karel-brinda , I am back at this for the paper review. First thing I noticed is that the pipeline is broken now, I can't run it. Upon further inspection, it is an error with the new prophyle version, `0.3.3.1`:

```
[Wed Oct  4 13:07:27 2023]
checkpoint prophyle_index:
    input: results/asm/test_run_with_tree/GCGS0001.fa, results/asm/test_run_with_tree/GCGS0002.fa, results/asm/test_run_with_tree/GCGS0003.fa, results/asm/test_run_with_tree/GCGS0004.fa, results/asm/test_run_with_tree/GCGS0005.fa, results/tree/test_run_with_tree.nw
    output: results/post/test_run_with_tree, results/post/test_run_with_tree/propagation
    jobid: 71
    reason: Missing output files: results/post/test_run_with_tree/propagation, results/post/test_run_with_tree
    wildcards: batch=test_run_with_tree
    resources: tmpdir=/tmp
DAG of jobs will be updated after completion.


        prophyle index  -T -A -S\
            -k 31 -g results/asm/test_run_with_tree\
            results/tree/test_run_with_tree.nw results/post/test_run_with_tree
        
Activating conda environment: .snakemake/conda/f6c1b60e36a2de96c30c74bb4f51dcfb_
[prophyle] 2023-10-04 13:07:28 Index construction started
[prophyle] 2023-10-04 13:07:28 Shell command: make -j -s -C /home/leandro/git/mof-compress/.test/.snakemake/conda/f6c1b60e36a2de96c30c74bb4f51dcfb_/lib/python3.10/site-packages/prophyle
/usr/bin/ld: /usr/bin/ld: DWARF error: can't find .debug_ranges section.
tree_index.o: in function `TreeIndex::fill_upper_nodes(knhx1_t const*)':
tree_index.cpp:(.text+0x5fe): undefined reference to `std::__throw_bad_array_new_length()'
/usr/bin/ld: tree_index.o: in function `std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash(unsigned long, unsigned long const&)':
tree_index.cpp:(.text._ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm[_ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm]+0xf7): undefined reference to `std::__throw_bad_array_new_length()'
/usr/bin/ld: tree_index.o: in function `std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash(unsigned long, unsigned long const&)':
tree_index.cpp:(.text._ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_S5_ESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm[_ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_S5_ESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm]+0xf7): undefined reference to `std::__throw_bad_array_new_length()'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:9: prophyle_assignment] Error 1
make: *** [Makefile:20: compile_assignment] Error 2
[prophyle] 2023-10-04 13:07:28 Unfinished, an error occurred (error code 2, 0.4 MB used): make -j -s -C /home/leandro/git/mof-compress/.test/.snakemake/conda/f6c1b60e36a2de96c30c74bb4f51dcfb_/lib/python3.10/site-packages/prophyle
ProPhyle Error: A command failed, see messages above.
[Wed Oct  4 13:07:28 2023]
Error in rule prophyle_index:
    jobid: 71
    input: results/asm/test_run_with_tree/GCGS0001.fa, results/asm/test_run_with_tree/GCGS0002.fa, results/asm/test_run_with_tree/GCGS0003.fa, results/asm/test_run_with_tree/GCGS0004.fa, results/asm/test_run_with_tree/GCGS0005.fa, results/tree/test_run_with_tree.nw
    output: results/post/test_run_with_tree, results/post/test_run_with_tree/propagation
    conda-env: /home/leandro/git/mof-compress/.test/.snakemake/conda/f6c1b60e36a2de96c30c74bb4f51dcfb_
    shell:
        
        prophyle index  -T -A -S\
            -k 31 -g results/asm/test_run_with_tree\
            results/tree/test_run_with_tree.nw results/post/test_run_with_tree
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Removing output files of failed job prophyle_index since they might be corrupted:
results/post/test_run_with_tree
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2023-10-04T130726.063716.snakemake.log
```

Downgrading it to the previous version, `0.3.2.0`, works just fine. So in the conda envs I will switch from using `>=` to `==` to prevent future updates potentially breaking the pipeline

karel-brinda commented:
Weird. This is related to C++ libraries and LD. Likely can be fixed by running `make clean` and `make`  in the corresponding directory of the C++ program. (I guess is reuses some files compiled elsewhere.)

In the future, this should be solved by moving ProphEx and ProphAsm out of ProPhyle and adding them as dependencies, but unfortunately don't have bandwidth for this (will require some correspoding updates in the scripts).

leoisl commented:
If it helps, this was done on a clean snakemake env, i.e. I removed the `.snakemake` dir before running the pipeline, and when I ran it, it had to recreate all the conda envs, which it did successfully, but then the pipeline failed when running the `prophyle_index` rule

karel-brinda commented:
This wouldn't help – these files come from Bioconda. It likely just requires running `make -C {DIR} clean` and `make {DIR}` where DIR is `/home/leandro/git/mof-compress/.test/.snakemake/conda/f6c1b60e36a2de96c30c74bb4f51dcfb_/lib/python3.10/site-packages/prophyle` (probably – you'd need to have a look)

leoisl commented:
Is it ok to just downgrade to the previous version for now? It is also more consistent with paper results, as they were done on the previous version

karel-brinda commented:
Btw. ProPhyle is necessary only if k-mer propagation is to be run. In the default conf, this shouldn't be done. Is it then possible to skip ProPhyle installation?

leoisl commented:
> If I recall correctly, at some point I added a new parameter to `prophyle index` that allows to skip FM-index construction (time consuming, but completely unnecessary in this case). Do you see if this is the difference between the 2 versions?

IDK, I don't know prophyle well and the new version is not running, so I can't compare both outputs

> Btw. ProPhyle is necessary only if k-mer propagation is to be run. In the default conf, this shouldn't be done. Is it then possible to skip ProPhyle installation?

would we then need two conda envs, one with and one without prophyle, and we activate the corresponding conda env depending what is in `config.yaml`? But anyway, installing prophyle is not an issue, is fast and it installs without issue. But when we try to run it, the new version fails. When we run prophyle, it seems it tries to compile stuff? Is there a way to do this compilation during its installation, not during running it?

karel-brinda commented:
ok, yse the old prophyle version and we will fix this later (we shouldn't get stuck by this)

leoisl commented:
Hey @karel-brinda , I've fixed package versioning in the conda envs and fixed the CI/CD pipeline. This pipeline is now running fine on ubuntu and Mac with python versions 3.8, 3.9, 3.10 and 3.11 (see the passed checks).

Just to refresh, as it has been some time since we were working on this PR, the main purpose of this PR is to make this pipeline configurable. Basically the user has to simply configure this `config.yaml`:
```
########################################################################################################################
# input configuration
# This parameter denotes the input dir, a path to a directory with a very specific structure, that contains the input
#     data to this pipeline, i.e. the batches.
# Here we describe this structure in details.
# Each batch named <batch> is composed of one required file (<batch>.txt) and one optional file (<batch>.nw).
#   * The <batch.txt> file is a text file where each line is a path to an assembly file;
#      * An assembly file is a file with extensions ["fa", "fasta", "fna", "ffa"], optionally compressed with gzip
#          ("gz" extension) denoting the assembly of a sample;
#      * Every *.txt file in the input dir will be interpreted as a batch;
#      * See .test/resources/test_run_with_tree.txt for an example of a <batch>.txt file;
#   * The <batch>.nw is an optional newick file representing the tree of the samples in its respective <batch>.txt
#          file. If this file is not provided, this pipeline builds a tree automatically using Mashtree.
#   * These files must be at the top level of the given directory.
input_dir: resources
########################################################################################################################


########################################################################################################################
# output configuration
output_dir: results
generate_assembly_stats: True  # True or False
generate_prepropagation_stats: True  # True or False
generate_postpropagation_stats: True  # True or False
########################################################################################################################
```

There is still a couple of stuff on my list:
1. Update the readme (in the style MOF-Search);
2. Lint the code and add linting to the CI pipeline;

However, these are quick to do and will be done in a PR posterior to this one. Then we can chat again and see what else we should improve
